.github/workflows/publish.yml

```diff
@@ -0,0 +1,112 @@
+name: Publish gem
+
+on: workflow_dispatch
+
+concurrency: "rubygems" # Only one publish job at a time
+
+jobs:
+  verify-checks:
+    name: Verify commit status checks
+    runs-on: ubuntu-24.04
+    permissions:
+      checks: read
+    outputs:
+      version: ${{ steps.version.outputs.version }}
+    steps:
+      - uses: actions/checkout@v4
+      - uses: ruby/setup-ruby@v1
+        with:
+          ruby-version: '3.3.7'
+
+      - id: version
+        run: echo "version=$(ruby -e 'puts Gem::Specification::load(Dir.glob("*.gemspec").first).version')" >> $GITHUB_OUTPUT
+
+      # Check if the gem version is already published
+      - name: Verify gem version
+        env:
+          GEM_VERSION: ${{ steps.version.outputs.version }}
+        run: |
+          if gem search datadog --exact --remote --version "$GEM_VERSION" | grep -q "($GEM_VERSION)"; then
+            echo "Version $GEM_VERSION is already published"
+            exit 1
+          else
+            echo "Version $GEM_VERSION is not published yet"
+          fi
+
+      # Check if the commit has passed all Github checks
+      - name: Verify check runs
+        uses: actions/github-script@v7
+        with:
+          script: |
+            const { data: checksResponse } = await github.rest.checks.listForRef({
```

Comment:
iirc there is a limit on the number of checks that can be returned by the rest API, even with pagination it'll only do the first 1k.

if there is a chance we'd ever have more this, the approach to use is to grab all of the check suites first, and then for each suite grab the list of checks for that suite.

Reply:
Would this get covered by the next check for the combined status?

Reply:
[Pagination example](https://github.com/DataDog/dd-trace-rb/blob/09169113cba7cf3656271db5d813ff39cb191cf9/.github/workflows/ensure-changelog-entry.yml#L42-L50) from another workflow

Reply:
>Would this get covered by the next check for the combined status?
Nope. These are different signals.

Pagination is addressed with: https://github.com/DataDog/dd-trace-rb/pull/4315/commits/de8d2c7dbb868053ef3144b3bef07050aeeb3f0a

Reply:
> Nope. These are different signals.

What is the difference? It is not clear, worth documenting or at least linking to REST api docs for them?

Reply:
>What is the difference?

Check runs are generated by Github Actions and commit statuses are provided externally. 

Add API endpoints doc in comment: 0842909242a52c8e9e3be1cc85ca4ce7698febf5.

Reply:
GitHub Actions don't also show up as commit statuses? that seems like a mistake on their part 🤦🏻

Reply:
https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/collaborating-on-repositories-with-code-quality-features/about-status-checks#types-of-status-checks-on-github
>There are two types of status checks on GitHub:
>- Checks
>- Commit statuses
> 
> Checks are different from commit statuses in that they provide line annotations, more detailed messaging, and are only available for use with GitHub Apps.

I think it is Github's design choice. Those statuses are means for developers to integrate with external providers. 

For example:  You can use the REST API to allow external services to mark commits with an error, failure, pending, or success state, which is then reflected in pull requests involving those commits